pkg/eventrecorder/eventrecorder.go

```diff
@@ -11,55 +11,127 @@ import (
 
 	"github.com/filecoin-project/lassie/pkg/events"
 	"github.com/filecoin-project/lassie/pkg/types"
-	logging "github.com/ipfs/go-log/v2"
-	"github.com/multiformats/go-multicodec"
+	"github.com/ipfs/go-log/v2"
 )
 
-var HttpTimeout = 5 * time.Second
-var ParallelPosters = 5
+var logging = log.Logger("eventrecorder")
 
-var log = logging.Logger("eventrecorder")
+const httpTimeout = 5 * time.Second
+const parallelPosters = 5
 
-type EventRecorderConfig struct {
-	DisableIndexerEvents  bool
-	InstanceID            string
-	EndpointURL           string
-	EndpointAuthorization string
+type data struct {
+	source string
+	event  types.RetrievalEvent
 }
 
-// NewEventRecorder creates a new event recorder with the ID of this instance
-// and the URL to POST to
-func NewEventRecorder(ctx context.Context, cfg EventRecorderConfig) *EventRecorder {
-	er := &EventRecorder{
-		ctx,
-		make(chan report),
-		make(chan []report),
-		cfg,
-	}
-
-	go er.ingestReports()
-	for i := 0; i < ParallelPosters; i++ {
-		go er.postReports()
+// NewEventRecorderSubsciber returns a RetrievalEventSubscriber that POSTs
+// a batch of retrieval events to an event recorder API endpoint
+func NewEventRecorderSubscriber(
+	ctx context.Context,
+	instanceID string,
+	endpointURL string,
+	endpointAuthorization string,
+	disableIndexerEvents bool,
+) types.RetrievalEventSubscriber {
+
+	incomingChan := make(chan data)
+	postChan := make(chan []data)
+
+	// Ingest data
+	go func() {
+		var batchedData []data
+		var batcherChan chan []data = nil
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			// read incoming data
+			case data := <-incomingChan:
+				batchedData = append(batchedData, data)
+				if batcherChan == nil {
+					batcherChan = postChan
+				}
+			case batcherChan <- batchedData:
```

Comment:
So are you just relying on a random diceroll for this here? batching based on which of these two gets picked, with the first being necessarily picked when `batcherChan` is `nil`? So doesn't that mean you're only going to get, on average, 1.5 events batched coming out the other end?

Reply:
This is the current behavior of the current EventRecorder, and was mimicked in the aggregated event recorder. This one is just less obviously named. And yes, this does result in batches of 1-3. I wasn't seeing batches of anything higher in AWS when reviewing Lambda logs.